scipy/stats/_sensitivity_analysis.py

```diff
@@ -0,0 +1,642 @@
+from __future__ import annotations
+
+import inspect
+from dataclasses import dataclass
+from typing import (
+    Callable, Dict, List, Literal, Protocol, TYPE_CHECKING, Tuple
+)
+
+import numpy as np
+
+from scipy.stats._common import ConfidenceInterval
+from scipy.stats._qmc import check_random_state
+from scipy.stats._resampling import BootstrapResult
+from scipy.stats import qmc, bootstrap
+
+
+if TYPE_CHECKING:
+    import numpy.typing as npt
+    from scipy._lib._util import DecimalNumber, IntNumber, SeedType
+
+
+__all__ = [
+    'sobol_indices',
+    'f_ishigami'
+]
+
+
+def f_ishigami(x: npt.ArrayLike) -> np.ndarray:
+    r"""Ishigami function.
+
+    .. math::
+
+        Y(\mathbf{x}) = \sin x_1 + 7 \sin^2 x_2 + 0.1 x_3^4 \sin x_1
+
+    with :math:`\mathbf{x} \in [-\pi, \pi]^3`.
+
+    Parameters
+    ----------
+    x : array_like ([x1, x2, x3], n)
+
+    Returns
+    -------
+    f : array_like (n,)
+        Function evaluation.
+
+    References
+    ----------
+    .. [1] Ishigami, T. and T. Homma. "An importance quantification technique
+       in uncertainty analysis for computer models." IEEE,
+       :doi:`10.1109/ISUMA.1990.151285`, 1990.
+    """
+    x = np.atleast_2d(x)
+    f_eval = (
+        np.sin(x[0])
+        + 7 * np.sin(x[1])**2
+        + 0.1 * (x[2]**4) * np.sin(x[0])
+    )
+    return f_eval
+
+
+def sample_A_B(
+    n: IntNumber,
+    dists: List[PPFDist],
+    random_state: SeedType = None
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Sample two matrices A and B.
+
+    Uses a Sobol' sequence with 2`d` columns to have 2 uncorrelated matrices.
+    This is more efficient than using 2 random draw of Sobol'.
+    See sec. 5 from [1]_.
+
+    Output shape is (d, n).
+
+    References
+    ----------
+    .. [1] Saltelli, A., P. Annoni, I. Azzini, F. Campolongo, M. Ratto, and
+       S. Tarantola. "Variance based sensitivity analysis of model
+       output. Design and estimator for the total sensitivity index."
+       Computer Physics Communications, 181(2):259-270,
+       :doi:`10.1016/j.cpc.2009.09.018`, 2010.
+    """
+    d = len(dists)
+    A_B = qmc.Sobol(d=2*d, seed=random_state, bits=64).random(n)
+
+    A, B = A_B[:, :d], A_B[:, d:]
+
+    for d_, dist in enumerate(dists):
+        try:
+            A[:, d_] = dist.ppf(A[:, d_])
+            B[:, d_] = dist.ppf(B[:, d_])
+        except AttributeError as exc:
+            message = (
+                "The method `ppf` must be specified for all distributions of "
+                "'dists'."
+            )
+            raise ValueError(message) from exc
+
+    return A.T, B.T
```

Comment:
No need to accept this, but I'd find this more readable. The ideas is that swapping the order of `try/except` and the loop is probably more efficient, and one call to `ppf` is better than two.

```suggestion
    A_B = qmc.Sobol(d=2*d, seed=random_state, bits=64).random(n).T
    A_B = A_B.reshape(2, d, -1)

    try:
        for d_, dist in enumerate(dists):
            A_B[:, d_] = dist.ppf(A_B[:, d_])
    except AttributeError as exc:
        message = "Each distribution in `dists` must have method `ppf`."
        raise ValueError(message) from exc

    return A_B
```
The rephrased message is a bit more concise, which has the added benefit of fitting on one line. If you accept the message suggestion, it would need to be changed in the test.

Reply:
The message is better 👍 

Makes sense to have a single call.
For the return value, what about we re-split at the end? `A, B = A_B[:, :d], A_B[:, d:]`. I would like to keep the output as 2 matrices. Or do you say we should just return a single matrix?

Reply:
Ok my bad it's all good!